include/circt/Dialect/FIRRTL/FIRRTLDeclarations.td

```diff
@@ -64,10 +64,11 @@ def MemOp : FIRRTLOp<"mem", [/*MemAlloc*/]> {
   let arguments = (ins I32Attr:$readLatency,
                        Confined<I32Attr, [IntMinValue<1>]>:$writeLatency,
                        I64Attr:$depth, RUWAttr:$ruw,
+                       StrArrayAttr:$portNames,
                        OptionalAttr<StrAttr>:$name);
-  let results = (outs BundleType:$result);
+  let results = (outs Variadic<FIRRTLType>:$results);
 
-  let assemblyFormat = "$ruw attr-dict `:` type($result)";
+  let assemblyFormat = "$ruw attr-dict (`:` type($results)^ )?";
```

Comment:
Is there ever going to be a memory with no ports?  Is that worth declaring invalid and hence results isn't optional?

Reply:
I expect that there will never be such a memory, but the SFC allows it and we have existing tests of such "empty" memories.

Such a memory could possibly exist if read/write ports are deleted due to some optimization, e.g., memory ports get DCE'd.

Reply:
Yeah, I ran into this as well.  We should accept it and canonicalize it away.

Reply:
Actually, let me argue the other way.  The alternative here is recognize zero ported mems in the parser, drop them, and verify they aren't possible.  I think that is is actually a better way to go (even though it differs from SFC) since there is at least one method that doesn't work with this degenerate case.  I'd support doing that, but it should be an independent patch.

Reply:
It's probably fine to drop it during parsing. It's impossible to connect to a memory without ports (i.e., a portless memory _is_ dead code). If some SFC custom transform is relying on this weird behavior, then we can implement that in a better way with a different transform (e.g., a port-less memory has ports dynamically added to it by a transform). 

Should this be changed to require a type on the memory here or in a follow-up? There's at least one test that checks that the port-less behavior is allowed and would need to be ripped out.

Reply:
Plz add the type requirement in the followup, that includes the parser change.  Thanks.

Reply:
Opened #531 to track this.